### ISSUE TITLE
UV Mapping?

### ISSUE THREAD

**Issue by SpicyMelonYT** [Scrawk/Marching-Cubes#10]
Hi, 

I found this project recently and find it really cool. I was looking to make something like this but your design is way better than what I was gonna do. However, I noticed that there is no UV mapping it seems. 

I am making a game "like minecraft" but I need each block to have its own texture. To do this I just need to generate the uv coords for each triangle. I actually managed to integrate this code well with mine and I have it generating a assumptive set of uv coords but they are not aligned correctly to the "quads" of each block since I am just aligning them the the triangles. So some textures are flipping around and not showing completely.

Is there any way you can implement correctly aligned quad uvs along with the verts and triangles. Or direct me on how I might go about it

I have this texture as a debug texture that I am trying to put on the blocks:

![Screenshot 2022-06-17 213026](https://user-images.githubusercontent.com/80491863/174422798-2c382b3e-45d4-481e-bc98-0337f19c5daa.png)

However this is the result with my method:
![Screenshot 2022-06-17 212732](https://user-images.githubusercontent.com/80491863/174422814-123ea63e-e300-42e8-89b6-8a7c5b22b538.png)

Here is the underlying geometry at that location (zoomed out but smaller image cause I suck at getting screen shots):
![Screenshot 2022-06-17 212912](https://user-images.githubusercontent.com/80491863/174422824-8e8ac245-a878-4384-aa57-532ab897dddf.png)

Ass you can see, the bottom right half of the texture is not being mapped because all the uvs are pulling from these coords: {x:0, y:0}, {x:1, y:0}, {x:1, y:1}. 

It seems there is nothing I can pull from to check if the current triangle is the top left triangle in the quad or the bottom right. 

This whole question might not be possible to solve, however, since many of the "quads" are skewed which means that the textures will be skewed. And if we were to attempt to place uvs in a way that made them not skewed, it would not tile along each "block"

Anyways thanks for the code!

**Comment by Scrawk** [comment]
There’s no way to generate uvs from the marching algorithm.

It’s something that’s going to custom to what look your trying to achieve.

The only method I can think of that might work is tri planar mapping.

**Comment by SpicyMelonYT** [comment]
I see, I will try a few more things still but thank you for the quick response!

**Comment by SpicyMelonYT** [comment]
OH I figured it out! Since the mesh that generates (no tetrahedron version) has all triangles only face 26 difference directions. I can take the vertex positions of each triangle corner and use that with a modulo operator to get all points mapped to the unit square of a texture. Then I can take that square and shrink it and offset it to place it onto the location in my texture atlas that corresponds to the type of block (dirt, stone, sand...etc) I got it working for top, right, and forward. I will have to do some mathy stuff to get it to "paste" correctly onto the angled surfaces but it should be the same concept. And as of right now it runs very well. Able to generate 256x256x256 chunks into my world in 3.23 seconds each if I wanted to. But 16x16x16 chunks are blazing fast so I will do it that way!
